amazon/ionbenchmark/self_describing_proto.py

```diff
@@ -0,0 +1,118 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from google.protobuf import message as _message
+from google.protobuf import message_factory as _message_factory
+from google.protobuf import reflection as _reflection
+
+from self_describing_proto_pb2 import SelfDescribingMessage
+
+import six
+
+
+class SelfDescribingProtoSerde:
+    """
+    This class provides functions for reading and writing self-describing protocol buffers.
+
+    This class uses the technique described in https://protobuf.dev/programming-guides/techniques/#self-description to
+    create a mostly self-describing protocol buffer. (It's not truly self-describing because the reader must know that
+    the serialized data uses the Self-describing Messages convention.)
+
+    To create a self-describing protocol buffer, see proto_tools.py.
+
+    This class makes no guarantees of thread-safety because each instance of this class has its own single instance of
+    the SelfDescribingMessage wrapper that it reuses for every load(s) call.
+
+    TODO: Add support for importing "well-known types" (https://protobuf.dev/reference/protobuf/google.protobuf/).
+    """
+    def __init__(self, cache_type_info=False, reuse_inner_object=False):
+        """
+        :param cache_type_info: Controls whether the type descriptor set and generated classes should be cached. Caching
+            the generated classes for load(s) can be significantly faster than not caching, but requires more memory for
+            the cache. WARNING: there is no cache eviction implemented yet.
+            (Has no effect for dump(s).)
+        :param reuse_inner_object: Controls whether loads should create a new instance of the inner message class each
+            time load(s) is called.
+            (Has no effect if cache_type_info is False. Has no effect for dump(s).)
+        """
+        self._cache_type_info = cache_type_info
+        self._reuse_inner_object = reuse_inner_object
+        self._cached_outer_object = SelfDescribingMessage()
+        self._cached_inner_definitions = {}
+        self._cached_inner_objects = {}
+
+    @staticmethod
+    def generate_class_definition(type_name, descriptor_set):
+        """
+        Generates a Python class for the given type_name using the provided descriptor_set.
+        """
+        messages_types = _message_factory.GetMessages(descriptor_set.file)
+        message_type = messages_types[type_name]()
+
+        class DynamicMessage(six.with_metaclass(_reflection.GeneratedProtocolMessageType, _message.Message)):
+            DESCRIPTOR = message_type.DESCRIPTOR
```

Comment:
Can you clarify if and how much other overhead the self-describing pattern adds? I'm mostly concerned that it's clear what we're benchmarking against: if most users are not using this pattern (as their doc implies) then is this going to give numbers that match reality?

Reply:
In the _very_ limited benchmarks I've done, the self-describing aspect adds enough overhead that it ends up being comparable to json. I plan to also add regular protobuf. I've included this because I think it's closer to a 1:1 comparison with Ion, and so I thought it would be useful. (It was also an easier first step for me to start integrating with the benchmark CLI since it doesn't require outside information in order to interpret the data.)

For ephemeral data (RPC messages) you would never use the self-describing pattern. However, if you were building a heterogeneous data lake, then the self-describing pattern would probably be a better fit.

Reply:
Thanks for the clarification. That makes sense and I second Tyler's insight:
> ... this self-describing format will demonstrate performance that Ion should be able to match in Ion 1.0.